src/Analyser/CoverageReporter.php

```diff
@@ -0,0 +1,457 @@
+use PHPStan\Broker\Broker;
+use PHPStan\File\FileExcluder;
+use PHPStan\File\FileHelper;
+use PHPStan\Parser\Parser;
+
+class CoverageReporter
+{
+
+	/**
+	 * @var \PHPStan\Parser\Parser
+	 */
+	private $parser;
+
+	/**
+	 * @var \PHPStan\Broker\Broker
+	 */
+	private $broker;
+
+	/**
+	 * @var \PHPStan\Analyser\NodeScopeResolver
+	 */
+	private $nodeScopeResolver;
+
+	/**
+	 * @var \PhpParser\PrettyPrinter\Standard
+	 */
+	private $printer;
+
+	/**
+	 * @var \PHPStan\Analyser\TypeSpecifier
+	 */
+	private $typeSpecifier;
+
+	/**
+	 * @var \PHPStan\File\FileExcluder
+	 */
+	private $fileExcluder;
+
+	/**
+	 * @var string|null
+	 */
+	private $bootstrapFile;
+
+	/**
+	 * @var bool
+	 */
+	private $reportUnmatchedIgnoredErrors;
+
+	/**
+	 * @param \PHPStan\Broker\Broker $broker
+	 * @param \PHPStan\Parser\Parser $parser
+	 * @param \PHPStan\Analyser\NodeScopeResolver $nodeScopeResolver
+	 * @param \PhpParser\PrettyPrinter\Standard $printer
+	 * @param \PHPStan\Analyser\TypeSpecifier $typeSpecifier
+	 * @param \PHPStan\File\FileExcluder $fileExcluder
+	 * @param \PHPStan\File\FileHelper $fileHelper
+	 * @param string|null $bootstrapFile
+	 */
+	public function __construct(
+		Broker $broker,
+		Parser $parser,
+		NodeScopeResolver $nodeScopeResolver,
+		\PhpParser\PrettyPrinter\Standard $printer,
+		TypeSpecifier $typeSpecifier,
+		FileExcluder $fileExcluder,
+		FileHelper $fileHelper,
+		string $bootstrapFile = null
+	)
+	{
+		$this->broker = $broker;
+		$this->parser = $parser;
+		$this->nodeScopeResolver = $nodeScopeResolver;
+		$this->printer = $printer;
+		$this->typeSpecifier = $typeSpecifier;
+		$this->fileExcluder = $fileExcluder;
+		$this->bootstrapFile = $bootstrapFile !== null ? $fileHelper->normalizePath($bootstrapFile) : null;
+	}
+
+	/**
+	 * @param string[] $files
+	 * @param \Closure|null $progressCallback
+	 * @return array
+	 */
+	public function analyse(array $files, \Closure $progressCallback = null): array
+	{
+		if ($this->bootstrapFile !== null) {
+			if (!is_file($this->bootstrapFile)) {
+				throw new \RuntimeException('Bootstrap file %s does not exist.', $this->bootstrapFile);
+			}
+
+			require_once $this->bootstrapFile;
+		}
+
+		$reports = [];
+		$this->nodeScopeResolver->setAnalysedFiles($files);
+		foreach ($files as $file) {
+			if ($this->fileExcluder->isExcludedFromAnalysing($file)) {
+				if ($progressCallback !== null) {
+					$progressCallback($file);
+				}
+
+				continue;
+			}
+
+			$typecheckWarnings = [];
+			$this->nodeScopeResolver->processNodes(
+				$this->parser->parseFile($file),
+				new Scope($this->broker, $this->printer, $this->typeSpecifier, $file),
+				function (\PhpParser\Node $node, Scope $scope) use (&$typecheckWarnings) {
+					if ($node instanceof \PhpParser\Node\Stmt\Trait_) {
+						return;
+					}
+
+					if ($node instanceof \PhpParser\Node\Expr\Variable) {
+						$typecheckWarnings = array_merge(
+							$typecheckWarnings,
+							$this->handleVariableNode($node, $scope)
+						);
```

Comment:
From performance point of view it's a lot better to do just `$typecheckWarnings[] =$this->handleVariableNode($node, $scope)` and flatten the array at the end.